### ISSUE TITLE
why use memory usage to calucate gc percent?

### ISSUE THREAD

**Issue by Dingshujie** [cch123/gogctuner#11]

**Comment by Dingshujie** [comment]
> > * Based on scavenging policy, go will not return released memory to os immediately, but nextgc use heapMarked, it diff with memory usage
> > * memory usage from /proc/ may be delay, when trigger gc, and may lead gc percent incorrectly.
> 
> reading from memstats heap is more accurate，but reading from memstats will cause stw，which is harmful
> 
> reading from proc is also not good，because some app has lots of goroutines，if goroutine stacks cost more memory than heap，we cannot get a proper new valve
> 
> if you have any good suggestion，im happy to accept

how about runtime/metrics? metrics like /gc/heap/goal:bytes ?

**Comment by cch123** [comment]
> 

this package is imported in Go 1.16, I'll consider about it

**Comment by cch123** [comment]
> > > * Based on scavenging policy, go will not return released memory to os immediately, but nextgc use heapMarked, it diff with memory usage
> > > * memory usage from /proc/ may be delay, when trigger gc, and may lead gc percent incorrectly.
> > 
> > 
> > reading from memstats heap is more accurate，but reading from memstats will cause stw，which is harmful
> > reading from proc is also not good，because some app has lots of goroutines，if goroutine stacks cost more memory than heap，we cannot get a proper new valve
> > if you have any good suggestion，im happy to accept
> 
> how about runtime/metrics? metrics like /gc/heap/goal:bytes ?

runtime/metrics is imported in Go 1.16, I'll consider about it

**Comment by Dingshujie** [comment]
now, 1.19.beta provide Soft memory limit

**Comment by cch123** [comment]
> now, 1.19.beta provide Soft memory limit

Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...

**Comment by cch123** [comment]
> now, 1.19.beta provide Soft memory limit

Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...

**Comment by Dingshujie** [comment]
> > now, 1.19.beta provide Soft memory limit
> 
> Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...
> 
> So I'll keep this lib as it is🤔

may be in same case，we can use gc tuner to reduce gc cycle time ?

**Comment by cch123** [comment]
> > > now, 1.19.beta provide Soft memory limit
> > 
> > 
> > Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...
> > So I'll keep this lib as it is🤔
> 
> ![image](https://user-images.githubusercontent.com/13744637/173273210-eab4085d-ce4d-4eab-bc14-f44c8acdfd1d.png) heap goal is min val among Limit and \gamma is equal to 1+\frac{GOGC}{100}
> 
> may be in same case，we can use gc tuner to reduce gc cycle time ?

I've read the proposal again, GOGC and GOMEMLIMIT will both affect the gc cycle and memory consumption, we need some experiment to know the exact effects... "useless" is a hasty conclusion.

**Comment by cch123** [comment]
> > > now, 1.19.beta provide Soft memory limit
> > 
> > 
> > Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...
> > So I'll keep this lib as it is🤔
> 
> ![image](https://user-images.githubusercontent.com/13744637/173273210-eab4085d-ce4d-4eab-bc14-f44c8acdfd1d.png) heap goal is min val among Limit and \gamma is equal to 1+\frac{GOGC}{100}
> 
> may be in same case，we can use gc tuner to reduce gc cycle time ?

<img width="1086" alt="image" src="https://user-images.githubusercontent.com/384546/173278965-dcc915eb-2dad-4fad-94a3-cca48993a0cd.png">

But I think setting GOGC to off will have a same effect as gc tuner?

**Comment by Dingshujie** [comment]
> > > > now, 1.19.beta provide Soft memory limit
> > > 
> > > 
> > > Yes, https://github.com/golang/proposal/blob/master/design/48409-soft-memory-limit.md, memory ballast and gc tuner will be useless when 1.19 finally released...
> > > So I'll keep this lib as it is🤔
> > 
> > 
> > ![image](https://user-images.githubusercontent.com/13744637/173273210-eab4085d-ce4d-4eab-bc14-f44c8acdfd1d.png) heap goal is min val among Limit and \gamma is equal to 1+\frac{GOGC}{100}
> > may be in same case，we can use gc tuner to reduce gc cycle time ?
> 
> <img alt="image" width="1086" src="https://user-images.githubusercontent.com/384546/173278965-dcc915eb-2dad-4fad-94a3-cca48993a0cd.png">
> 
> But I think setting GOGC to off will have a same effect as gc tuner?

proposal says deal with Death spirals is to place a limit on the amount of total CPU utilization of the garbage collector. If the garbage collector were to execute and exceed that limit at any point, it will instead let the application proceed, even if that means missing its goal and breaching the memory limit. and proposal says they simply pick a reasonable default based on GOGC. 

may be need to check implement of Soft memory limit to known real action.
